File: 프로그래머스/3/67259. ［카카오 인턴］ 경주로 건설/［카카오 인턴］ 경주로 건설.java
import java.util.Arrays;
import java.util.LinkedList;
import java.util.Queue;

class Solution {
    static int[] dx = {0, 1, 0, -1};
    static int[] dy = {1, 0, -1, 0};
    static int l;
    public int solution(int[][] board) {
        l = board.length;
        int[][][] cost = new int[l][l][4];
        for (int i = 0; i < l; i++) {
            for (int j = 0; j < l; j++) {
                Arrays.fill(cost[i][j], Integer.MAX_VALUE);
            }
        }
        Queue<int[]> q = new LinkedList<>();
        q.offer(new int[]{0,0,-1, 0});
        while(!q.isEmpty()) {
            int[] arr = q.poll();
            int cx = arr[0];
            int cy = arr[1];
            if(cx == l - 1 && cy == l - 1) continue;
            int dir = arr[2];
            int currentCost = arr[3];
            for(int i = 0; i < 4; i++) {
                int nx = cx + dx[i];
                int ny = cy + dy[i];
                if(!Check(nx, ny) || board[nx][ny] == 1) continue;

                int newCost = currentCost + 100;
                if(dir != i && dir != -1) newCost += 500;
                
                if(cost[nx][ny][i] > newCost) {
                    cost[nx][ny][i] = newCost;
                    q.offer(new int[]{nx,ny,i,newCost});
                }
            }
        }
        int answer = Integer.MAX_VALUE;
        for(int i = 0; i < 4; i++)
            answer = Math.min(answer ,cost[l-1][l-1][i]);
        return answer;
    }
    public static boolean Check(int nx, int ny) {
        return 0 <= nx && nx < l && 0 <= ny && ny < l;
    }
}

File: 프로그래머스/3/67259. ［카카오 인턴］ 경주로 건설/［카카오 인턴］ 경주로 건설.py
from collections import deque

def check(x,y,n): return 0 <= x < n and 0 <= y < n

def solution(board):
    # 우,하,좌,상 순
    directions = [(0,1),(1,0),(0,-1),(-1,0)]
    n = len(board)
    # 상하좌우로 해당 칸 입성할 때 최솟값
    visit = [[[float('inf')] * n for _ in range(n)] for _ in range(4)]
    q = deque()
    # 아래 방향
    # x,y,dir,cost
    q.append((0,0,1,0))
    # 오른쪽 방향
    q.append((0,0,0,0))
    visit[1][0][0] = 0
    visit[0][0][0] = 0
    while q:
        cx,cy,cd,cost = q.popleft()
        for i,d in enumerate(directions):
            nx,ny = cx + d[0], cy + d[1]
            if not check(nx,ny,n) or board[nx][ny] == 1: continue
            # 우 하 상 좌
            # 현재 방향과 일치 -> 100
            if cd == i or (cd+2) % 4 == i:
                # 최소 금액이 아니라면 pass
                if visit[i][nx][ny] <= cost + 100: continue
                visit[i][nx][ny] = cost + 100
                q.append((nx,ny,i,cost+100))
            # 코너 -> 600
            else:
                # 최소 금액 아니라면 pass
                if visit[i][nx][ny] <= cost + 600: continue
                visit[i][nx][ny] = cost + 600
                q.append((nx,ny,i,cost+600))
    ans = float('inf')
    for i in range(4):
        ans = min(ans,visit[i][n-1][n-1])
    return ans